File: 1026_Maximum_Difference_Between_Node_and_Ancestor.js
// Leetcode 1026 - Maximum Difference Between Node and Ancestor
// https://leetcode.com/problems/maximum-difference-between-node-and-ancestor/

/**
 * Definition for a binary tree node:
 * function TreeNode(val, left, right) {
 *     this.val = (val === undefined ? 0 : val);
 *     this.left = (left === undefined ? null : left);
 *     this.right = (right === undefined ? null : right);
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxAncestorDiff = function(root) {
    // Recursive DFS function that tracks min/max along the current path
    const dfs = (node, minv, maxv) => {
        if (!node) {
            // Return the max difference between ancestor and node
            return maxv - minv;
        }

        // Update current min and max with node value
        minv = Math.min(minv, node.val);
        maxv = Math.max(maxv, node.val);

        // Explore both subtrees
        const left = dfs(node.left, minv, maxv);
        const right = dfs(node.right, minv, maxv);

        // Return the max difference found
        return Math.max(left, right);
    };

    return dfs(root, root.val, root.val);
};

File: 1026_Maximum_Difference_Between_Node_and_Ancestor.py
# Leetcode 1026 - Maximum Difference Between Node and Ancestor
# https://leetcode.com/problems/maximum-difference-between-node-and-ancestor/

# The idea: At each node, track the minimum and maximum values from root to that node.
# The difference between max and min gives the max ancestor difference at that path.

class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        # Recursive DFS function
        def dfs(node, minv, maxv):
            if not node:
                # Return max difference when we reach a leaf
                return maxv - minv

            # Update min and max with the current node's value
            minv = min(minv, node.val)
            maxv = max(maxv, node.val)

            # Recur for left and right subtrees
            left = dfs(node.left, minv, maxv)
            right = dfs(node.right, minv, maxv)

            # Return the max diff found from both sides
            return max(left, right)

        # Start DFS from root, using its value as initial min and max
        return dfs(root, root.val, root.val)
